### ISSUE TITLE
Deprecate CJS in favour of ESM

### ISSUE THREAD

**Issue by thecodingdude** [nodejs/modules#46]
Hi folks - although I am not a member of this team, I am (and have been) following the progress of node implementing ESM for a long time now - it's been in the works since 2016 and after a few years it feels like it's always _so_ close but not quite there yet. I apologise if this discussion has already happened before, but I wanted to get the communities thoughts on potentially scrapping CJS in favour of ESM.

In order to save some time, I want to point to a very good [article](https://medium.com/@dlmanning/interoperability-between-es-modules-and-common-js-is-a-mistake-fb9ac71d96fd) written in 2016 discussing the interoperability between these two systems.

Reading that post, it appears there are two solutions that I can see:

1:

If you load something with import it is parsed as a Module,
If you load with require(), it is parsed as a Script

2:

Remove CJS altogether - maintain 2 NodeJS binaries (v10 LTS, v11, v12 LTS) which ships CJS or ESM only, deprecate CJS after v12 LTS.

Right now, interoperability between CJS and ESM does not exist and can not exist. The very fact Node has decided to use ".mjs" puts it as a second class citizen; the browser does not use .mjs, and will not use .mjs, the Node community will stick with .js and I am pretty sure most tooling will not be updated to work with .mjs.

You guys also need to think _very_ carefully about the repercussions of introducing the potential damage a new file extension may cause - 5, 10 years down the line, you will look back and realise it was a fundamental mistake - .js is _how_ Node works and has _always_ worked.

The reality is quite simple here: CJS is not and was not a standard. It existed before ESM was standardized and implemented in browsers and now that's the case, the _only_ logical progression is to remove it from Node entirely, supporting two module loading systems is going to be _so much pain_ that you are probably not even aware of. You'll only find out once you unprefix the flag and actually have _millions_ of developers using it. It makes no sense to keep them both.

Mozilla deprecated their _entire_ extension ecosystem in Firefox 57 after they rewrote it from scratch meaning _all_ old plugins were obsolete in favour of Web extensions.

AngularJS 1.x and Angular 2.x+ have coexisted for many years.

Python 2.7 and 3.x have coexisted for many years where major changes were made.

There is precedence for this and trying to support two different ways of loading a JS file inside another is just plain crazy, and will be a massive mistake if you even try to attempt that. 

The only thing as a developer that I request is you do _not_ unflag modules until there has been _extensive_ feedback from multiple parties including "regular" developers (maybe a githib emoji poll or something) to understand whether it's a plausible implementation. Obviously the potential damage this can cause is too great to simply release this without thought.

Thank you.

**Comment by streamich** [comment]
So you are saying Node should stick with only one module system. But why do you pick ESM over CJS?

**Comment by thecodingdude** [comment]
ESM is a [standard](https://www.ecma-international.org/ecma-262/6.0/#sec-imports), CJS is not.
ESM is being implemented in browsers, CJS is not.
CJS is NOT a standard, which makes Node officially non-standard complaint. 

Node, in theory, should try to ship standards complaint code where possible - trying to support two different module loaders feels redundant. ESM has replaced CJS and therefore should ship in Node by default, which would be far, far less work in the long run than trying to shoehorn .mjs which is a terrible idea and will see absolutely _no_ adoption. Many, many tools depend on ".js" and will not recognize ".mjs" and will create unnecessary fragmentation.

**Comment by devsnek** [comment]
@thecodingdude Thanks for taking the time to keep yourself reasonably informed (not many people do 😄). We did in fact actively discuss throwing out CJS and we agreed that was not the best path. As a team, we have committed to shipping ESM and CJS to provide paths for people to migrate. In the future we definitely want to see ESM being the default, and we are still discussing different ways for this to happen. All of us see `.mjs` as temporary, but it will also be necessary for making sure we provide the community with a way out of CJS. ESM in node isn't finished yet, and we're still probably going to be making a lot of changes.

Also just as a clarification:

> If you load something with import it is parsed as a Module,
If you load with require(), it is parsed as a Script

this isn't actually true in either case. require will load .json, .node files, (other things with require.extensions but you shouldn't use that) and `import` is specifically left vague so it may load basically anything e.g. wasm, cjs, images, html (basically anything with a proper loader exposed, which is what whatwg is working on [here](https://github.com/whatwg/loader))

**Comment by ljharb** [comment]
Scrapping CJS isn’t an option imo because the vast majority of npm - the largest package manager on the planet - is CJS. Back compat with the ecosystem is critical.

**Comment by thecodingdude** [comment]
@devsnek 

> As a team, we have committed to shipping ESM and CJS to provide paths for people to migrate

Yet you mention

> All of us see .mjs as temporary, but it will also be necessary

Which concerns me to an alarming degree - it could be said many people saw CJS as "temporary", and would've been replaced with ESM that only got finalized in 2015 and only got _implemented_ in Chrome in September 2017 with the unprefixed version due to come in Firefox 60 (not yet released).

There is no such thing as "temporary" in programming, and especially node. Once you unflag modules with ".mjs" you are _committed_ to that path - tooling gets updated, IDE's become aware of .mjs and it becomes a "thing" that will stick around and haunt you if you decide it's only for a few years.

Rather than trying to get modules in v10, or v12 or v15 even, it'd be better for _everyone_ if there was an absolute concrete plan in how ESM and CJS can coexist in a way that makes sense, that's easy to migrate to and will work out in the long term.

This issue is designed to raise a discussion on how we can eventually remove CJS and replace it in favour of ESM - and the truth is, .mjs is _not_ the answer, and it will be a _disaster_ for node in another 5-10 years, when you one day realise "oh shit...we can't deprecate this now". So you're stuck with CJS and ESM that uses a file extension nobody wanted. James Snell described it as a 'compromise' in a medium post, where there are no such things that are affordable in such a widely used language; we should strive to get this _right_ first time and not create future problems that you _know_ you'll need to deal with.

@ljharb - yet, time and again we see as I listed in my examples that this is the _only_ way platforms can grow and evolve over time. Mozilla didn't decide to random break ALL their extensions because "lulz"; they are investing in the next 10 years and whilst it's painful _today_, in 5 years time _everyone_ will be grateful for what they did.

People will migrate from Python 2, or AngularJS 1.x, because they _have_ to, because one day, support will stop. And you can either use broken, insecure code, or get with the times and use a vastly improved software which is designed for _today_. This insistence that software cannot _ever_ make breaking changes is insanity and needs removing from peoples arbitrary 'book of rules'.   

All I want to do is discuss _how_ we can either make CJS exist with ESM using .js, or find a way to remove CJS and make ESM a first class citizen, which is entirely possible and doesn't necessitate .mjs, there have been many great discussions about how we can do this, for example, [NPM's suggestion](https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054) seems to make sense and not introduce the ridiciousness that is .mjs.

**Comment by devsnek** [comment]
tl;dr: npm's proposal is something we are discussing and we didn't randomly decide on mjs because "lulz"

thank you for your concern about where node is going but please believe us when we say we're working on making node the best it can be. i'm going to close this since it just seems like its turning into another complaint piece on mjs.
